Guidance on ABI stability (#34)
pmix/governance: pmix_governance.tex

the changes will be handled using the full rigorous process
as described in Section~\ref{the-standardization-process}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\hypertarget{abi-changes-guidance}{%
\subsubsection{Application Binary Interface (ABI) Changes Guidance}%
\label{abi-changes-guidance}}

The PMIx Standard should strive to maintain a stable \acf{ABI} to support applications and tools that rely on more than one implementation of the PMIx Standard.
To facilitate such interoperability the PMIx \ac{ASC} will maintain, to the best of their ability, a standard set of headers that are versioned with the PMIx Standard that applications and tools can reference~\footnote{PMIx Headers for ABI Compatibility \url{https://github.com/pmix/pmix-abi}}.
In recognition that there are circumstances where the \ac{ABI} needs to be modified this section defines some guidance for making such modifications.
Additions to the PMIx interface can occur without breaking \ac{ABI} compatibility.

> Removing portions of the PMIx interface are covered in Section~\ref{deprecation} on Deprecation and Removal and follow these guidelines.

Author: ```suggestion
Removing portions of the PMIx interface are covered in Section~\ref{deprecation} on Deprecation and Removal and follow those guidelines.
```

As such the following text focuses on modifications to the existing PMIx interface.

The PMIx Standard must not modify the \ac{ABI} in minor releases.
All modifications of the \ac{ABI} must occur in a major release.
If a Provisional (see Section~\ref{provisional-acceptance}) change requires a modification of the \ac{ABI} then it is no longer eligible to be included in a minor release and must wait to be included in the next major release.

> Where possible data structures should use padding where appropriate to reduce the impact of adding fields to the data structure on the stability of the \ac{ABI}.

Author: ```suggestion
When adding a field to an existing data structure that has padding then that padding must be adjusted so as to preserve the stability of the \ac{ABI}.
```

the changes will be handled using the full rigorous process
as described in Section~\ref{the-standardization-process}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\hypertarget{abi-changes-guidance}{%
\subsubsection{Application Binary Interface (ABI) Changes Guidance}%
\label{abi-changes-guidance}}

An \acf{API} defines how data types, data structures, and functions are represented in source code.
An \acf{ABI} defines how data types, data structures, and functions are represented in machine code for a given system.
An important aspect of an \ac{ABI} is the size, layout, and alignment of data structures.
A stable \ac{ABI} may allow a program compiled with one implementation of the PMIx Standard to run with a different implementation of the PMIx Standard as long as both implementations adhere to the same \ac{ABI}.

The PMIx Standard should strive to maintain a stable \ac{ABI} to support applications and tools that rely on more than one implementation of the PMIx Standard.

> To facilitate such interoperability the PMIx \ac{ASC} will maintain, to the best of their ability, a standard set of headers that are versioned with the PMIx Standard that applications and tools can reference~\footnote{PMIx Headers for ABI Compatibility \url{https://github.com/pmix/pmix-abi}}.

Author: ```suggestion
To facilitate such interoperability the PMIx \ac{ASC} will maintain a standard set of headers that are versioned in relation to the PMIx Standard document that applications and tools can reference~\footnote{PMIx Headers for ABI Compatibility \url{https://github.com/pmix/pmix-abi}}.
```

Author: Addressed in the newest commit.

> 

Author: Add the following:
-----
The PMIx Standard ABI is defined in two parts. The Stable ABI represents the Stable PMIx Standard elements (see Section 1.3.1 of the PMIx Governance Document). The Provisional ABI represents the Provisional PMIx Standard elements (see Section 1.3.1 of the PMIx Governance Document). The rationale for this separation is in recognition that Provisional PMIx Standard elements are able to change frequently in response to broad experimentation. This is in contrast to Stable PMIx Standard elements which are permanent, non-changing parts of the PMIx Standard.

Both the Stable and Provisional ABIs are versioned with two increasing numbers (largely following the Semantic Versioning 2.0.0 specification [1]):
 * `MAJOR` incremented when the ABI changed in a backwards incompatable manner.
 * `MINOR` incremented when functionality is added to the ABI in a backwards compatible manner.

The PMIx ASC will maintain a mapping between the PMIx Standard version and the ABI version.

[1] https://semver.org/spec/v2.0.0.html

Author: Addressed in the newest commit.

In recognition that there are circumstances where the \ac{ABI} needs to be modified this section defines some guidance for making such modifications.
Additions to the PMIx interface can occur without breaking \ac{ABI} compatibility.
Deprecating and removing portions of the PMIx interface does break \ac{ABI} compatibility and that process (see Section~\ref{deprecation}) must continue to adhere to the following guidance.
As such the following text focuses on modifications to the existing PMIx interface.

The PMIx Standard must not modify the \ac{ABI} in minor releases.
All modifications of the \ac{ABI} must occur in a major release.

> If a Provisional (see Section~\ref{provisional-acceptance}) change requires a modification of the \ac{ABI} then it is no longer eligible to be included in a minor release and must wait to be included in the next major release.

Author: Since we now have a Provisional ABI version, do we want to allow Provisional change to change their ABI in minor versions? It seems like we do since it will continue to allow Provisional items to move quickly.
```suggestion
```

Author: Note from the IAWG Meeting on April 18, 2022:
 * Stable ABI changes
   * `MAJOR` increments only happen in major releases of the PMIx Standard
   * `MINOR` increments can happen in either major or minor releases of the PMIx Standard
 * Provisional ABI changes
   * `MAJOR` increments can happen in either major or minor releases of the PMIx Standard
   * `MINOR` increments can happen in either major or minor releases of the PMIx Standard
 * Rationale: This preserves the spirit of Provisional changes to be a place for experimentation.

Author: Addressed in the newest commit.
